tmk_core/common/chibios/eeprom_stm32.c

```diff
@@ -48,72 +58,124 @@ uint16_t EEPROM_Init(void) {
-    page = FEE_ADDR_OFFSET(Address) / FEE_PAGE_SIZE;
+// we are sure the address will not be out of bound
+#pragma GCC diagnostic ignored "-Warray-bounds"
+    // same value, do not need program
+    if (DataBuf[Address] == DataByte) {
+        return FlashStatus;
+    }
 
-    // if current data is 0xFF, the byte is empty, just overwrite with the new one
-    if ((*(__IO uint16_t *)(FEE_PAGE_BASE_ADDRESS + FEE_ADDR_OFFSET(Address))) == FEE_EMPTY_WORD) {
-        FlashStatus = FLASH_ProgramHalfWord(FEE_PAGE_BASE_ADDRESS + FEE_ADDR_OFFSET(Address), (uint16_t)(0x00FF & DataByte));
+    // find the address can be written
+    addr = EEPROM_FindValidAddress();
+    DataBuf[Address] = DataByte;
+#pragma GCC diagnostic pop
+    if (addr == FEE_EMPTY_VALUE) {
+        // EEPROM is full, need to erase and reprogramming
+        EEPROM_Clear();
+        EEPROM_Restore();
     } else {
-        // Copy Page to a buffer
-        memcpy(DataBuf, (uint8_t *)FEE_PAGE_BASE_ADDRESS + (page * FEE_PAGE_SIZE), FEE_PAGE_SIZE);  // !!! Calculate base address for the desired page
-
-        // check if new data is differ to current data, return if not, proceed if yes
-        if (DataByte == *(__IO uint8_t *)(FEE_PAGE_BASE_ADDRESS + FEE_ADDR_OFFSET(Address))) {
-            return 0;
-        }
-
-        // manipulate desired data byte in temp data array if new byte is differ to the current
-        DataBuf[FEE_ADDR_OFFSET(Address) % FEE_PAGE_SIZE] = DataByte;
-
-        // Erase Page
-        FlashStatus = FLASH_ErasePage(FEE_PAGE_BASE_ADDRESS + (page * FEE_PAGE_SIZE));
-
-        // Write new data (whole page) to flash if data has been changed
-        for (i = 0; i < (FEE_PAGE_SIZE / 2); i++) {
-            if ((__IO uint16_t)(0xFF00 | DataBuf[FEE_ADDR_OFFSET(i)]) != 0xFFFF) {
-                FlashStatus = FLASH_ProgramHalfWord((FEE_PAGE_BASE_ADDRESS + (page * FEE_PAGE_SIZE)) + (i * 2), (uint16_t)(0xFF00 | DataBuf[FEE_ADDR_OFFSET(i)]));
-            }
-        }
+        FLASH_ProgramHalfWord(FEE_ADDR_ADDRESS(addr), Address);
+        FLASH_ProgramHalfWord(FEE_DATA_ADDRESS(addr), DataByte);
     }
+
     return FlashStatus;
 }
 /*****************************************************************************
  *  Read once data byte from a specified address.
  *******************************************************************************/
 uint8_t EEPROM_ReadDataByte(uint16_t Address) {
-    uint8_t DataByte = 0xFF;
+    if (!IS_VALID_ADDRESS(Address)) {
+        return FEE_EMPTY_BYTE;
+    }
 
-    // Get Byte from specified address
-    DataByte = (*(__IO uint8_t *)(FEE_PAGE_BASE_ADDRESS + FEE_ADDR_OFFSET(Address)));
+    // Get Byte from caches
+    return DataBuf[Address];
+}
 
-    return DataByte;
+/*****************************************************************************
+ *  helper functions
+ *******************************************************************************/
+// backup the current data
+void EEPROM_Backup(void)
```

Comment:
suggest renaming to `EEPROM_Load` or similar, Backup doesn't really make sense to me as it's loaded once during initialization